src/BlueprintParser.h

```diff
@@ -383,43 +406,98 @@ namespace snowcrash {
+                initialDeps = pd.namedTypeDependencyTable[identifier];
+                diffDeps.clear();
+
+                for (std::set<mson::Literal>::iterator it = initialDeps.begin();
+                     it != initialDeps.end();
+                     it++) {
+
+                    std::set<mson::Literal> superTypeDeps = pd.namedTypeDependencyTable[*it];
+                    pd.namedTypeDependencyTable[identifier].insert(superTypeDeps.begin(), superTypeDeps.end());
+                }
+
+                // Check if the list of dependents has grown
+                finalDeps = pd.namedTypeDependencyTable[identifier];
+                std::set_difference(finalDeps.begin(), finalDeps.end(), initialDeps.begin(), initialDeps.end(),
+                                    std::inserter(diffDeps, diffDeps.end()));
+
+            } while (!diffDeps.empty());
+        };
+
+        /**
          * \brief For each entry in the named type inheritance table, resolve the sub-type's base type recursively
          *
          * \param pd Section parser data
          * \param subType The sub named type between the two
          * \param superType The super named type between the two
          * \param report Parse report
-         * \param inheritanceNodes Set containing named types which have been checked for inheritance in current cycle
          */
-        static void resolveNamedTypeTableEntry(SectionParserData& pd,
-                                               const mson::Literal& subType,
-                                               const mson::Literal& superType,
-                                               const mdp::BytesRangeSet& nodeSourceMap,
-                                               Report& report,
-                                               std::set<mson::Literal> inheritanceNodes) {
+        static void resolveNamedTypeBaseTableEntry(SectionParserData& pd,
+                                                   const mson::Literal& subType,
+                                                   const mson::Literal& superType,
+                                                   const mdp::BytesRangeSet& nodeSourceMap,
+                                                   Report& report) {
 
             mson::BaseType baseType;
             mson::NamedTypeBaseTable::iterator it = pd.namedTypeBaseTable.find(subType);
 
-            inheritanceNodes.insert(superType);
-
             // If the base table entry is already filled, nothing else to do
             if (it != pd.namedTypeBaseTable.end()) {
                 return;
             }
 
+            // Check for circular references
+            std::set<mson::Literal> deps = pd.namedTypeDependencyTable[subType];
+
+            if (deps.find(subType) != deps.end()) {
+
+                // ERR: A named type is circularly referenced
+                std::stringstream ss;
+                ss << "base type '" << subType << "' circularly referencing itself is not allowed";
```

Comment:
Not sure where this is exactly being printed out. But I would say something like "recursive base type XXXX" , I would definitely skip the "is not allowed" part